[bnd-baseline-maven-plugin 3.2.0] Kura API baseline verification fails

Hello,
  I'm trying to use the bnd-baseline-maven-plugin to verify the Kura API baseline.
I've run the analysis on this [branch](https://github.com/eclipse/kura/tree/enable-bnd-baseline-maven-plugin)

The verification fails for some API packages (see below). Full maven log [here](https://s3.amazonaws.com/kura-issues-logs/bnd-baseline-maven-plugin.log.zip).

For example, if we consider the `org.eclipse.kura.configuration` package exported by the Kura `org.eclipse.kura.api` bundle, I believe we properly changed the MINOR version from [1.0.0](https://github.com/eclipse/kura/blob/KURA_1.4.0_RELEASE/kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/configuration/ConfigurationService.java) to [1.1.0](https://github.com/eclipse/kura/blob/enable-bnd-baseline-maven-plugin/kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/configuration/ConfigurationService.java), having just added some methods and a constant to the `org.eclipse.kura.configuration.CloudService` interface. Changes to other classes in the same package do not touch the API.

If I'm not wrong, Semantic Versioning only requires a change to the MINOR version.
However the plugin complains that:

```
[ERROR] Baseline mismatch for package org.eclipse.kura.configuration, MAJOR change. Current is 1.1.0, repo is 1.0.0, suggest 2.0.0 or 1.0.0
```

I don't understand why we should change the MAJOR number (2.0.0) and why the plugin suggest (1.0.0) as an alternative.

```
# attrs for org.eclipse.kura version="1.2.0"
# attrs for org.eclipse.kura.bluetooth version="1.2.0"
# attrs for org.eclipse.kura.certificate version="1.0.0"
# attrs for org.eclipse.kura.clock version="1.0.0"
# attrs for org.eclipse.kura.cloud version="1.0.0"
# attrs for org.eclipse.kura.cloud.factory version="1.0.0"
# attrs for org.eclipse.kura.comm version="1.0.0"
# attrs for org.eclipse.kura.command version="1.1.0"
# attrs for org.eclipse.kura.configuration version="1.1.0"
# attrs for org.eclipse.kura.configuration.metatype version="1.0.0"
# attrs for org.eclipse.kura.crypto version="1.2.0"
# attrs for org.eclipse.kura.data version="1.1.0"
# attrs for org.eclipse.kura.data.listener version="1.0.0"
# attrs for org.eclipse.kura.data.transport.listener version="1.0.0"
# attrs for org.eclipse.kura.db version="1.0.0"
# attrs for org.eclipse.kura.gpio version="1.0.0"
# attrs for org.eclipse.kura.linux.udev version="1.0.0"
# attrs for org.eclipse.kura.message version="1.0.0"
# attrs for org.eclipse.kura.net version="1.2.0"
# attrs for org.eclipse.kura.net.dhcp version="1.0.0"
# attrs for org.eclipse.kura.net.dns version="1.0.0"
# attrs for org.eclipse.kura.net.firewall version="1.0.0"
# attrs for org.eclipse.kura.net.modem version="1.1.0"
# attrs for org.eclipse.kura.net.route version="1.0.0"
# attrs for org.eclipse.kura.net.wifi version="1.2.0"
# attrs for org.eclipse.kura.position version="1.1.0"
# attrs for org.eclipse.kura.security version="1.0.0"
# attrs for org.eclipse.kura.ssl version="1.2.0"
# attrs for org.eclipse.kura.status version="1.0.0"
# attrs for org.eclipse.kura.system version="1.1.0"
# attrs for org.eclipse.kura.usb version="1.1.0"
# attrs for org.eclipse.kura.watchdog version="1.0.0"
[ERROR] Baseline mismatch for package org.eclipse.kura.bluetooth, MAJOR change. Current is 1.2.0, repo is 1.1.0, suggest 2.0.0 or 1.2.0

[ERROR] Baseline mismatch for package org.eclipse.kura.configuration, MAJOR change. Current is 1.1.0, repo is 1.0.0, suggest 2.0.0 or 1.0.0

[ERROR] Baseline mismatch for package org.eclipse.kura.data, MAJOR change. Current is 1.1.0, repo is 1.0.0, suggest 2.0.0 or 1.0.0

[ERROR] Baseline mismatch for package org.eclipse.kura.net.wifi, MINOR change. Current is 1.2.0, repo is 1.2.0, suggest 1.3.0 or -

[ERROR] Baseline mismatch for package org.eclipse.kura.position, MAJOR change. Current is 1.1.0, repo is 1.0.0, suggest 2.0.0 or 1.1.0
```

Hello Cristiano,

The version change required by Semantic versioning is different depending on the use of the API.

For example, adding a method to javax.servlet.ServletContext is a minor change, it only breaks the people who implement a Servlet Container. Adding a method to javax.servlet.Servlet is a major change, as it breaks all people who write servlets.

In this case the baselining plugin has determined that the org.eclipse.kura.configuration.CloudService interface is a “consumer type” i.e. it is implemented by other bundles that would be considered “clients” of the API. If this determination is correct then the change genuinely is a breaking API change, and needs to be a 2.0.0. If this is not the case then you should mark the interface as @ProviderType using the OSGi annotations. That way bnd will know which versioning policy to use, and 1.1.0 will be acceptable.

I’m afraid that I can’t answer why it’s suggesting 1.0.0 as an alternative, except to say that that would be the correct version if you reverted the API changes.

Tim Ward

Apache Aries PMC Member,
OSGi IoT Expert Group Chair,
Author Enterprise OSGi in Action
timothyjward@apache.org<mailto:timothyjward@apache.org>





On 15 Jul 2016, at 11:45, Cristiano De Alti <notifications@github.com<mailto:notifications@github.com>> wrote:


Hello,
I'm trying to use the bnd-baseline-maven-plugin to verify the Kura API baseline.
I've run the analysis on this branch<https://github.com/eclipse/kura/tree/enable-bnd-baseline-maven-plugin>

The verification fails for some API packages (see below). Full maven log here<https://s3.amazonaws.com/kura-issues-logs/bnd-baseline-maven-plugin.log.zip>.

For example, if we consider the org.eclipse.kura.configuration package exported by the Kura org.eclipse.kura.api bundle, I believe we properly changed the MINOR version from 1.0.0<https://github.com/eclipse/kura/blob/KURA_1.4.0_RELEASE/kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/configuration/ConfigurationService.java> to 1.1.0<https://github.com/eclipse/kura/blob/enable-bnd-baseline-maven-plugin/kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/configuration/ConfigurationService.java>, having just added some methods and a constant to the org.eclipse.kura.configuration.CloudService interface. Changes to other classes in the same package do not touch the API.

If I'm not wrong, Semantic Versioning only requires a change to the MINOR version.
However the plugin complains that:

[ERROR] Baseline mismatch for package org.eclipse.kura.configuration, MAJOR change. Current is 1.1.0, repo is 1.0.0, suggest 2.0.0 or 1.0.0


I don't understand why we should change the MAJOR number (2.0.0) and why the plugin suggest (1.0.0) as an alternative.

# attrs for org.eclipse.kura version="1.2.0"
# attrs for org.eclipse.kura.bluetooth version="1.2.0"
# attrs for org.eclipse.kura.certificate version="1.0.0"
# attrs for org.eclipse.kura.clock version="1.0.0"
# attrs for org.eclipse.kura.cloud version="1.0.0"
# attrs for org.eclipse.kura.cloud.factory version="1.0.0"
# attrs for org.eclipse.kura.comm version="1.0.0"
# attrs for org.eclipse.kura.command version="1.1.0"
# attrs for org.eclipse.kura.configuration version="1.1.0"
# attrs for org.eclipse.kura.configuration.metatype version="1.0.0"
# attrs for org.eclipse.kura.crypto version="1.2.0"
# attrs for org.eclipse.kura.data version="1.1.0"
# attrs for org.eclipse.kura.data.listener version="1.0.0"
# attrs for org.eclipse.kura.data.transport.listener version="1.0.0"
# attrs for org.eclipse.kura.db version="1.0.0"
# attrs for org.eclipse.kura.gpio version="1.0.0"
# attrs for org.eclipse.kura.linux.udev version="1.0.0"
# attrs for org.eclipse.kura.message version="1.0.0"
# attrs for org.eclipse.kura.net<http://org.eclipse.kura.net> version="1.2.0"
# attrs for org.eclipse.kura.net.dhcp version="1.0.0"
# attrs for org.eclipse.kura.net.dns version="1.0.0"
# attrs for org.eclipse.kura.net.firewall version="1.0.0"
# attrs for org.eclipse.kura.net.modem version="1.1.0"
# attrs for org.eclipse.kura.net.route version="1.0.0"
# attrs for org.eclipse.kura.net.wifi version="1.2.0"
# attrs for org.eclipse.kura.position version="1.1.0"
# attrs for org.eclipse.kura.security version="1.0.0"
# attrs for org.eclipse.kura.ssl version="1.2.0"
# attrs for org.eclipse.kura.status version="1.0.0"
# attrs for org.eclipse.kura.system version="1.1.0"
# attrs for org.eclipse.kura.usb version="1.1.0"
# attrs for org.eclipse.kura.watchdog version="1.0.0"
[ERROR] Baseline mismatch for package org.eclipse.kura.bluetooth, MAJOR change. Current is 1.2.0, repo is 1.1.0, suggest 2.0.0 or 1.2.0

[ERROR] Baseline mismatch for package org.eclipse.kura.configuration, MAJOR change. Current is 1.1.0, repo is 1.0.0, suggest 2.0.0 or 1.0.0

[ERROR] Baseline mismatch for package org.eclipse.kura.data, MAJOR change. Current is 1.1.0, repo is 1.0.0, suggest 2.0.0 or 1.0.0

[ERROR] Baseline mismatch for package org.eclipse.kura.net.wifi, MINOR change. Current is 1.2.0, repo is 1.2.0, suggest 1.3.0 or -

[ERROR] Baseline mismatch for package org.eclipse.kura.position, MAJOR change. Current is 1.1.0, repo is 1.0.0, suggest 2.0.0 or 1.1.0


—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/bndtools/bnd/issues/1553>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ADs8eE1VDbntTr6zO9A0hgqenWrKEP6Hks5qV2THgaJpZM4JNR-w>.

Can we reopen this?

I'm still struggling to have this working and merged in our Eclipse Kura build.
I've properly annotated some of the interfaces of the ``org.eclipse.kura.api bundle`` in [my Kura fork](https://github.com/cdealti/kura/tree/api-analysis/kura/org.eclipse.kura.api) with ``@ProviderType``.

I've set the baseline to the released ``org.eclipse.kura.api bundle`` [v.1.1.0](https://github.com/eclipse/kura/tree/KURA_2.1.0_RELEASE/kura/org.eclipse.kura.api). Please note that the baseline is not yet annotated with ``@ProviderType`` so the first question is if the analysis is likely to fail in this case. I had the API analysis working in the past, on a previous baseline, so my impression was that the plugin did not really care if the baseline is annotated or not.

```
Downloading: https://repo.eclipse.org/content/repositories/kura-releases/org/eclipse/kura/org.eclipse.kura.api/1.1.0/org.eclipse.kura.api-1.1.0.jar
Downloaded: https://repo.eclipse.org/content/repositories/kura-releases/org/eclipse/kura/org.eclipse.kura.api/1.1.0/org.eclipse.kura.api-1.1.0.jar (198 KB at 98.1 KB/sec)

[ERROR] Baseline mismatch for package org.eclipse.kura, MAJOR change. Current is 1.2.1, repo is 1.2.1, suggest 2.0.0 or -
...
[ERROR] Baseline mismatch for package org.eclipse.kura.gpio, MAJOR change. Current is 1.0.1, repo is 1.0.1, suggest 2.0.0 or -
...
[ERROR] The bundle version change (1.1.0 to 1.2.0) is too low, the new version must be at least 2.0.0
```

As we can see the plugin thinks that I have two major changes.

The first one is for the package ``org.eclipse.kura``. This package only contains various Kura Exception **classes**, which has changed since the baseline - not sure if an incompatible way, and no **interfaces**. Hence I haven't annotated them. The first question is whether the plugin analyses these classes and what type of binary compatibility analysis it does. Semantically the exception classes in this package should be ``@noextend``.

The second package ``org.eclipse.kura.gpio`` hasn't changed since the baseline. I've only added @ProviderType on two interfaces but this should result in a CHANGED (micro version) change.

Any idea of what's going on?

--- a/maven/bnd-baseline-maven-plugin/src/main/java/aQute/bnd/maven/baseline/plugin/BaselineMojo.java
+++ b/maven/bnd-baseline-maven-plugin/src/main/java/aQute/bnd/maven/baseline/plugin/BaselineMojo.java
@@ -33,6 +33,8 @@ import aQute.bnd.differ.Baseline.BundleInfo;
 import aQute.bnd.differ.Baseline.Info;
 import aQute.bnd.differ.DiffPluginImpl;
 import aQute.bnd.osgi.Jar;
+import aQute.bnd.service.diff.Delta;
+import aQute.bnd.service.diff.Diff;
 import aQute.libg.reporter.ReporterAdapter;
 import aQute.service.reporter.Reporter;
 
@@ -206,6 +208,20 @@ public class BaselineMojo extends AbstractMojo {
 				logger.error("Baseline mismatch for package {}, {} change. Current is {}, repo is {}, suggest {} or {}",
 						info.packageName, info.packageDiff.getDelta(), info.newerVersion, info.olderVersion,
 						info.suggestedVersion, info.suggestedIfProviders == null ? "-" : info.suggestedIfProviders);
+				if (fullReport) {
+					logger.info(info.packageDiff.toString());
+					for (Diff typeDiff : info.packageDiff.getChildren()) {
+						logger.info("    " + typeDiff.toString());
+						for (Diff fieldOrMethodDiff : typeDiff.getChildren()) {
+							// Unchanged field/method diffs are noisy and
+							// irrelevant
+							if (fieldOrMethodDiff.getDelta() == Delta.UNCHANGED) {
+								continue;
+							}
+							logger.info("        " + fieldOrMethodDiff.toString());
+						}
+					}
+				}
 			}
 		}
 

